server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java

```diff
@@ -241,24 +265,36 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
 
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
+        periodicRechecker.cancel();
         if (event.localNodeMaster()) {
             if (shouldReassignPersistentTasks(event)) {
-                logger.trace("checking task reassignment for cluster state {}", event.state().getVersion());
-                clusterService.submitStateUpdateTask("reassign persistent tasks", new ClusterStateUpdateTask() {
-                    @Override
-                    public ClusterState execute(ClusterState currentState) {
-                        return reassignTasks(currentState);
-                    }
-
-                    @Override
-                    public void onFailure(String source, Exception e) {
-                        logger.warn("failed to reassign persistent tasks", e);
-                    }
-                });
+                reassignPersistentTasks(event.state().getVersion());
+            } else {
+                periodicRechecker.schedule();
             }
         }
     }
 
+    /**
+     * Submit a cluster state update to reassign any persistent tasks that need reassigning
+     */
+    private void reassignPersistentTasks(long currentStateVersion) {
+        logger.trace("checking task reassignment for cluster state {}", currentStateVersion);
+        clusterService.submitStateUpdateTask("reassign persistent tasks", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                ClusterState newState = reassignTasks(currentState);
+                periodicRechecker.schedule();
```

Comment:
Also, this should be done when the cluster state is published. See `ClusterStateUpdateTask.clusterStatePublished`

Reply:
`ClusterStateUpdateTask.clusterStatePublished` is empty and `final` and also only gets called if the cluster state is updated by the `execute` method.

I think to make this reschedule after publishing or no-op I'd need to change the update class to implement`AckedClusterStateTaskListener` right?  Then `onAllNodesAcked` will get called regardless of whether `execute` changes the cluster state or not.

Any objections before I start on this change?

Reply:
argh, you are right. The one you need is `clusterStateProcessed` . Note that both of these are only called when the cluster state is committed, which is a big difference than when the execute method returns. It means that master got other master nodes to agree on the change.

Reply:
It looks like overriding `clusterStateProcessed` is non-trivial now.

Since https://github.com/elastic/elasticsearch/pull/31241/files#diff-bc0dd060947fa9d8e3209d60f7255f1dR67 it cannot be overridden in a class that extends `ClusterStateUpdateTask` because it runs in the system context.  I don't think running in the system context would be a problem for persistent task allocation decisions, but it creates a question of how to do this without opening up `ClusterStateUpdateTask` to future mistakes.  I could move most of the functionality of `ClusterStateUpdateTask` into a new base class called `SystemContextClusterStateUpdateTask`, and have `ClusterStateUpdateTask` extend that and just override `clusterStateProcessed` with it's empty `final` version.  Does that sound OK?

Reply:
@droberts195 I may be missing something, but the way I read the other PR, it is exactly the intention to use `clusterStateProcessed` as the thread context is maintained there. See [here](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/cluster/service/MasterService.java#L477).

Reply:
Sorry, I was still looking at `clusterStatePublished`.  `clusterStateProcessed` makes it easy.